aten/src/ATen/native/SoftMax.cpp

```diff
@@ -50,6 +51,60 @@ TORCH_META_FUNC(_log_softmax) (
 
   set_output(input.sizes(), output_options);
 }
+
+TORCH_META_FUNC(_softmax_backward_data)
+(const Tensor& grad,
+ const Tensor& output,
+ int64_t dim,
+ const Tensor& input) {
+  TensorArg grad_arg{grad, "grad", 1}, output_arg{output, "output", 2};
+  checkSameSize("softmax_backward", grad_arg, output_arg);
+
+  int64_t dim_ = maybe_wrap_dim(dim, grad.dim());
+
+  auto grad_input_options =
+      grad.options().memory_format(LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+
+  bool half_to_float = grad.scalar_type() != input.scalar_type();
+  if (half_to_float) {
+    if (grad.scalar_type() == ScalarType::Float &&
+         input.scalar_type() == ScalarType::Half) {
+      grad_input_options = grad_input_options.dtype(ScalarType::Half);
+    }
+  }
```

Comment:
So this snippet appeared only in the CUDA implementaiton, but CPU forward `softmax` and `log_softmax` kernels do not support `half_to_float=true`. Interestingly, such a check is not present in `softmax_backward` and `log_softmax_backward`, but those ops would only run if their forward counterparts were used and didn't crash, right? However, for some reason (as I learned from a failing test), the expression `grad.scalar_type() != input.scalar_type()` is `true` for `softmax_backward` even though its forward counterpart does not support `half_to_float=true`. So that means the `TORCH_CHECK` has to stay in the CUDA code while the output sizing happens here.

I'm not comfortable with this but tests are passing

Reply:
So just to confirm what you are saying, you used to have TORCH_CHECK for half to float in the meta function https://github.com/pytorch/pytorch/pull/62372/files/f8e92384918a1c3eb9034a0d671a196c9cf53308#diff-fa020f1f5fbaaade906225ece79758e020492f144e596338ad6ad9e693059657R68-R76 but this failed this test:

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 378, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 373, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 2496, in wrapped
    f(self, *args, **kwargs, coalesced=True)
  File "test_sparse.py", line 3204, in test_softmax
    test_op(1, 10, [2, 3], coalesced)
  File "test_sparse.py", line 3170, in test_op
    J3 = softmax_jacobian_autograd(x, dim)
  File "test_sparse.py", line 3122, in softmax_jacobian_autograd
    y.backward(v)
  File "/opt/conda/lib/python3.6/site-packages/torch/_tensor.py", line 274, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 156, in backward
    allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
RuntimeError: expected input and grad types to match, or input to be at::Half and grad to be at::Float
```

Reply:
I do agree this is a bit unsavory. Looking over the kernels, there is no reference to the actual data in input; the casting behavior appears to be about ensuring that the output gradient matches dtype of input (which is very reasonable; in fact, in an ideal world you would probably set the meta function directly based on the input options (but I bet, as the code is, this will break things). I looked over the test case and I couldn't figure out why there was a mismatch in the input/output dtype (but it's definitely not half precision related; probably something between double and float). I agree the code that you have here is semantics preserving.

I think it's probably worth writing a note about this, with a reference to the relevant failing test; e.g., file an issue for someone to follow up on later.

Reply:
Definitely want a comment inline in the code here, 'dis shit is weird yoooo

Reply:
I added a comment about this and filed #63057.